lib/web/app/transport.go

```diff
@@ -135,6 +135,12 @@ func newTransport(c *transportConfig) (*transport, error) {
 // RoundTrip will rewrite the request, forward the request to the target
 // application, emit an event to the audit log, then rewrite the response.
 func (t *transport) RoundTrip(r *http.Request) (*http.Response, error) {
+	// Clone the request so we can modify it without affecting the original.
+	// This is necessary because the request cookies are deleted when the web
+	// handler forward the request to the app proxy. When this happens, the
+	// cookies are lost and the error handler will not be able to find the
+	// session based on cookies.
+	r = r.Clone(r.Context())
```

Comment:
and while we're all looking at this code, I have a question for everyone:

in `lib/web/app.Handler.handleForwardError` we attempt to renew the session and re-send the request if the error is a `trace.ConnectionProblem`.

As far as I can tell, the only way that error is returned is if the transport's `DialContext` func has no app servers to try dialing: https://github.com/gravitational/teleport/blob/977e0136bf9862aa6135ea3155a9c2d1a23d30db/lib/web/app/transport.go#L272

So presumably, this retrying will only succeed if/when app servers become available, otherwise it will just keep calling the error forwarder over and over.

But even when it does succeed to dial an app server, the request body will have been closed by the initial call to `RoundTrip` - as required by the `RoundTripper` interface contract:

> **RoundTrip must always close the body, including on errors**

I verified that the `http.Transport` indeed closes the request body on a dial error.

So any request that gets into this code path will:
1. keep retrying until an app server is found, if ever
2. already have a closed request body on retry

Is there anything wrong in this assessment? Should we remove this retry behavior?

Reply:
I plan to fix this in a separate PR because it's the current behavior of Teleport - it requires saving the read data besides handling the close part.

Reply:
> Does it close the client connection? If it is, then the retry won't work anymore.

it only closes the body